### PR TITLE
Implement relational NULL semantics for Nothing for in-memory Column operations

### PR DESCRIPTION
**Pull request by GregoryTravis**
### Pull Request Description

Updates in-memory table column operations to treat Nothing as a relational NULL.
This PR does not include changes to Table.join.
### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Column.enso [hunk 1]
```diff
@@ -1600,7 +1600,7 @@ type Column
                 run_vectorized_binary_op self op_name as_vector expected_result_type=Value_Type.Boolean skip_nulls=False new_name=result_name
             False ->
                 set = Set.from_vector as_vector error_on_duplicates=False
-                run_unary_op self set.contains new_name=result_name skip_nulls=False expected_result_type=Value_Type.Boolean
+                run_unary_op self set.contains_relational new_name=result_name skip_nulls=True expected_result_type=Value_Type.Boolean
```
Comment (GregoryTravis) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Data/Column.enso]: @radeusgd I'm not sure how to invoke the fallback implementation of `==` (line 214) -- what case uses this?

Comment (radeusgd) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Data/Column.enso]: I need more context for this question. Line 214 of which file?

File: std-bits/table/src/main/java/org/enso/table/data/column/operation/map/SpecializedIsInOp.java [hunk 2]
```diff
@@ -55,16 +55,21 @@ public Storage<?> runMap(S storage, List<?> arg) {
     Context context = Context.getCurrent();
     CompactRepresentation<T> compactRepresentation = prepareList(arg);
     BitSet newVals = new BitSet();
+    BitSet missing = new BitSet();
     for (int i = 0; i < storage.size(); i++) {
-      if (storage.isNa(i) && compactRepresentation.hasNulls) {
-        newVals.set(i);
+      if (storage.isNa(i)) {
+        missing.set(i);
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/data/column/operation/map/SpecializedIsInOp.java]: What if `arg.size() == 0`? Looking at the `contains_relational` semantics, it should result in `False`, but looking at this code it seems that it will result in `Nothing`.

Do we have a test case for such scenario? If not, please let's add it and ensure the implementation is correct.

File: std-bits/table/src/main/java/org/enso/table/util/ImmutableBitSet.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
+package org.enso.table.util;
+
+import java.util.BitSet;
+
+/**
+ * A wrapper around BitSet that implements boolean operations conveniently. Unlike BitSet,
+ * ImmutableBitSet takes a size parameter, which allows .not to be implemented.
+ */
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/util/ImmutableBitSet.java:8]: Cool!

File: test/Table_Tests/src/Common_Table_Operations/Nothing_Spec.enso [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+            + [[False, False, False, False, False]]
+            + [[Nothing, True, True, True, Nothing]]
+            + [[Nothing, True, False, True, Nothing]]
+            + [[Nothing, False, True, True, Nothing]]
+            + [[Nothing, False, False, True, Nothing]]
+            + [[Nothing, True, True, False, Nothing]]
+            + [[Nothing, True, False, False, Nothing]]
+            + [[Nothing, False, True, False, Nothing]]
+            + [[Nothing, False, False, False, False]]
+
```
Comment (radeusgd) [inline test/Table_Tests/src/Common_Table_Operations/Nothing_Spec.enso]: IMO we should generate this programmatically, it is very hard to parse and check this huge table for correctness.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Column.enso (hunk 1 above)
Comment (GregoryTravis) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Data/Column.enso]: Line 214 of this file -- I can't leave a comment there.
Or here: https://github.com/enso-org/enso/blob/develop/distribution/lib/Standard/Table/0.0.0-dev/src/Data/Column.enso#L214

I think this code might need to be modified for handling nulls differently, but I'm not sure what cases use it.

Comment (radeusgd) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Data/Column.enso]: Try using `==` on a `Mixed` column, e.g.
```
t = Table.new [["X", ["a", 42, Nothing]], ["Y", ["b", 42, True]]]
(t.at "X") == (t.at "Y")
```

I think it should trigger it, but you can add some println to make sure :)

File: std-bits/table/src/main/java/org/enso/table/data/column/operation/map/SpecializedIsInOp.java (hunk 2 above)
Comment (GregoryTravis) [reply std-bits/table/src/main/java/org/enso/table/data/column/operation/map/SpecializedIsInOp.java]: Done.

File: test/Table_Tests/src/Common_Table_Operations/Nothing_Spec.enso (hunk 4 above)
Comment (GregoryTravis) [reply test/Table_Tests/src/Common_Table_Operations/Nothing_Spec.enso]: Done.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Column.enso (hunk 1 above)
Comment (GregoryTravis) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Data/Column.enso]: Ah, I remember now -- `==` uses skip_nulls=True, so the fallback is never used for nulls. So no change is needed.

Comment (radeusgd) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Data/Column.enso]: Oh :D

File: std-bits/table/src/main/java/org/enso/table/data/column/operation/map/bool/BooleanIsInOp.java [hunk 5]
```diff
@@ -60,43 +60,30 @@ public Storage<?> runZip(
   }
 
   private BoolStorage run(BoolStorage storage, boolean hadNull, boolean hadTrue, boolean hadFalse) {
```
Comment (AdRiley) [inline std-bits/table/src/main/java/org/enso/table/data/column/operation/map/bool/BooleanIsInOp.java:62]: Perhaps just my lack of knowledge of the codebase, but what is the difference between Null, Nothing and Missing? This PR is about **Nothing**, this function takes a had**Null** and generates a new**Missing**.

File: std-bits/table/src/main/java/org/enso/table/util/ImmutableBitSet.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    assert size == other.size;
+    BitSet result = (BitSet) bitSet.clone();
+    result.flip(0, size);
+    result.and(other.bitSet);
+    result.flip(0, size);
+    return new ImmutableBitSet(result, size);
+  }
+
+  public static ImmutableBitSet allFalse(int size) {
+    return new ImmutableBitSet(new BitSet(), size);
```
Comment (AdRiley) [inline std-bits/table/src/main/java/org/enso/table/util/ImmutableBitSet.java:77]: Is 

return new ImmutableBitSet(new BitSet(size), size);

going to make the allocation more efficient?

File: std-bits/table/src/main/java/org/enso/table/util/ImmutableBitSet.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+  }
+
+  public static ImmutableBitSet allTrue(int size) {
+    return new ImmutableBitSet(new BitSet(), size).not();
```
Comment (AdRiley) [inline std-bits/table/src/main/java/org/enso/table/util/ImmutableBitSet.java:81]: Similarly

return new ImmutableBitSet(new BitSet(size), size).not(); ?

File: test/Table_Tests/src/Common_Table_Operations/Nothing_Spec.enso [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+            _ : Database_Column -> vorc.not                     # Set negated bit
+        negation_cases = [["normal input, normal argument", identity, identity]]
+            + [["negated input, normal argument", negate, identity]]
+            + [["normal input, negated argument", identity, negate]]
+            + [["negated input, negated argument", negate, negate]]
+
+        negation_cases.map negation_case->
+            negation_desc = negation_case.at 0
+            neg_input = negation_case.at 1
+            neg_argument = negation_case.at 2
```
Comment (radeusgd) [inline test/Table_Tests/src/Common_Table_Operations/Nothing_Spec.enso]: nitpick but IMO this is misleading `neg_input` may be `negate` but it may as well be `identity`. So the `neg_` is misleading what this is.

Maybe just `transform_input` instead?

```suggestion
            transform_input = negation_case.at 1
            transform_argument = negation_case.at 2
```

**Review by radeusgd — APPROVED** [review]

File: std-bits/table/src/main/java/org/enso/table/data/column/operation/map/bool/BooleanIsInOp.java (hunk 5 above)
Comment (radeusgd) [reply std-bits/table/src/main/java/org/enso/table/data/column/operation/map/bool/BooleanIsInOp.java:62]: I think they all mean the same concept here really, so it is just a bit of chaos that got into the codebase over time.

- `Nothing` is the Enso-centrict name.
- `Null` probably came from the Java internals naming convention (where the `Nothing` is often represented as a `null` reference).
- _missing_ is the naming that stemmed from also old version of Table (we used to have `Column.is_missing` instead of `Column.is_nothing`; the external API got renamed, but the internals likely stayed. The `*Missing` bitsets are used when we are storing values like integers in unboxed `long[]` arrays - such don't take `null` (like `Object[]` could), so there needs to be a separate flag of which values are missing. I guess it could be called `isNothing` though, for consistency.

We could probably rename all of these to `nothing`-based names.

**Review by jdunkerley — APPROVED** [review]

File: std-bits/table/src/main/java/org/enso/table/util/ImmutableBitSet.java (hunk 6 above)
Comment (radeusgd) [reply std-bits/table/src/main/java/org/enso/table/util/ImmutableBitSet.java:77]: Looking at the usages, I think to the contrary.

The thing is that `new BitSet()` creates an empty bitset, which will answer `false` to any `bitset.get(ix)`.

So it creates the `allFalse` bitset while using the minimal amount of memory. Whereas specifying the `size`, will pre-allocate the memory - which in this case just seems unnecessary.

So actually I think that keeping _not_ adding the `size` is best for allocation (but I'm only 85% sure about it).

**Review by AdRiley — APPROVED** [review]

File: std-bits/table/src/main/java/org/enso/table/util/ImmutableBitSet.java (hunk 6 above)
Comment (GregoryTravis) [reply std-bits/table/src/main/java/org/enso/table/util/ImmutableBitSet.java:77]: Yes, `BitSet` is really a set of 1-positions, not a set of boolean values, which is why I needed `ImmutableBitSet` to have a size field, so it would really be more like a set of bits. Specifically, you can't use `.not()` unless you know what the last bit position is. (The `not`-related methods are the only ones that use `size` for something other than checking that the size are the same.)

File: test/Table_Tests/src/Common_Table_Operations/Nothing_Spec.enso (hunk 8 above)
Comment (GregoryTravis) [reply test/Table_Tests/src/Common_Table_Operations/Nothing_Spec.enso]: Done.

File: std-bits/table/src/main/java/org/enso/table/util/ImmutableBitSet.java (hunk 3 above)
Comment (JaroslavTulach) [reply std-bits/table/src/main/java/org/enso/table/util/ImmutableBitSet.java:8]: - What a wonderful class
- It could be very useful for #14409
